include/swift/AST/DiagnosticsSema.def

```diff
@@ -142,8 +142,20 @@ ERROR(cannot_pass_rvalue_mutating_getter,none,
       "cannot use mutating getter on immutable value of type %0",
       (Type))
 
-ERROR(expression_too_complex,none,
-      "the compiler is unable to type-check this expression in reasonable time; "
+ERROR(expression_too_complex_hard_memory,none,
+      "the compiler is unable to type-check this expression in reasonable time. Solver memory limit reached; "
+      "try breaking up the expression into distinct sub-expressions", ())
+ERROR(expression_too_complex_hard_time,none,
+      "the compiler is unable to type-check this expression in reasonable time. Solver time limit reached; "
+      "try breaking up the expression into distinct sub-expressions", ())
+ERROR(expression_too_complex_predicted_exponential_ratio,none,
+      "the compiler is unable to type-check this expression in reasonable time. Predicting solver memory exhaustion. Heuristic: type assignments/disjunctions; "
+      "try breaking up the expression into distinct sub-expressions", ())
+ERROR(expression_too_complex_predicted_exponential_disjunction_count,none,
+      "the compiler is unable to type-check this expression in reasonable time. Predicting solver memory exhaustion. Heuristic: high disjunction count; "
+      "try breaking up the expression into distinct sub-expressions", ())
+ERROR(expression_too_complex_default,none,
```

Comment:
These are good for a debug build, but I fail to see how this improves the end-user experience.

Reply:
Take memory limits, disjunction counts, or stack depth in the solver for example: We’re trading an opaque error for a slightly more specific opaque error that uses compiler jargon.

Reply:
I'll respond in more detail later today, but I was going to say effectively the same. These are no more actionable than what we have today, and are now tied to heuristics which we know aren't ideal. Further, the advice someone gives on <insert your favorite Q&A site> for one expression with one of these messages won't necessarily apply to the expression of another person who is hitting the same message.

Reply:
I concur with @rudkx and @CodaFi  on this one, ideally we shouldn't have this error at all but if we do we should probably avoid mentioning any implementation details and make it as general as possible.